### ISSUE TITLE
PollingClient and OpenStream

### ISSUE THREAD

**Issue by warappa** [NEventStore/NEventStore#387]
Currently it's not possible to use the PollingClient and open a stream in a subscription-action. An "DataReader already open"-exception is thrown.

My scenario:
I need the checkpoint-tokens of events on the *command-side* (to check somewhere else with latest readside checkpoint), so I use PollingClient (on the command-side) to get the commit's `CheckpointToken`.
Then I route the events (still on the *command-side*), so *command-handlers* waiting for specific events ("saga-light") can check the eventstore for an streamId (and in my case for a different aggregate-type on the same aggregate-id as my repository-implementation supports that) and, if not found, create it.

Currently this is broken for me. Previously i used the async dispatcher - which is now outdated...

A naive fix, I think, would be to do a `ToArray()` on the commits in
https://github.com/NEventStore/NEventStore/blob/78d994dc8bf1269132bc2dd8ab282eabf01a44e4/src/NEventStore/Client/PollingClient.cs#L115
but then we might run out of memory some day.

Another approach would be batching, only getting, say, a maximum of 100 commits at a time like 
```
IEnumerable<ICommit> commits = _persistStreams.GetFrom(_checkpointToken).Take(100).ToArray();
```
handle that and then do the foreach ... subject.Next(...) ... _checkpointToken = ... stuff.

            private void DoPoll()
            {
                if (Interlocked.CompareExchange(ref _isPolling, 1, 0) == 0)
                {
                    try
                    {
                        while (true)
                        {
                            IEnumerable<ICommit> commits = _persistStreams.GetFrom(_checkpointToken)
                                .Take(100)
                                .ToArray();

                            if (commits.Count() == 0)
                                  break;

                            foreach (var commit in commits)
                            {
                                if (_stopRequested.IsCancellationRequested)
                                {
                                    _subject.OnCompleted();
                                    return;
                                }
                                _subject.OnNext(commit);
                                _checkpointToken = commit.CheckpointToken;
                            }
                        }
                    }
                    catch (Exception ex)
                    {
                        // These exceptions are expected to be transient
                        Logger.Error(ex.ToString());
                    }
                    Interlocked.Exchange(ref _isPolling, 0);
                }
            }

What do you think?

**Comment by warappa** [comment]
As a workaround I tried to use 2 independent eventstores - same connection-string, different connection-names. But I still got the exception!

I debugged it and it seems that the cause lies in `ConfigurableConnectionFactory` in connection with `ConnectionScope`. `ConfigurableConnectionFactory.Open(connectionName)` creates a new `ConnectionScope`. `ConnectionScope` has as 1st parameter *`connectionName`* but the connection-*string* is passed. Then `ThreadScope` uses that connection-string as key in the *cache* - and that's the problem:
2 different named connection-strings using equal connection-strings therefore *share the same connection!* I think that is not right.

https://github.com/NEventStore/NEventStore/blob/master/src/NEventStore/Persistence/Sql/ConfigurationConnectionFactory.cs#L59

EDIT:
Just tried adding a single semicolon at the end of the 2nd connection-string - now it works!

**Comment by andreabalducci** [comment]
To get the commit's checkpoint for the command you could override the PostCommit method on a custom PipelineHook.

**Comment by warappa** [comment]
(Sorry for the delay)

First of all: Thanks for your reply!

Replacing `PollingClient` with a pipeline-hook would be possible, generally speaking, but I store all the commands and command-results (failed, succeeded) in 2 streams: **`commands` and `command-result` streams**. After that the command is stored, it's loaded from the store, handled and it's result stored in the results-stream.
Yes, async command handling...

Well, if the app dies after persisting the command prior to handling them, I would still have to catch-up with the outstanding commands with some method (`PollingClient.ObserveFrom(checkpoint)` or `Store.Advanced.GetFrom(checkpoint)`) and only after that start using the pipeline-hook. To me it seems that it's easier to just use `PollingClient` all the time and that's it.

(Of course there is the problem with how to get the last handled checkpoint but that's another issue...)

@ Original issue:
I will setup a small pull-request to the CommandFactory to prevent an (in my opinion) accidental reuse of a connection if 2 connection-strings have the same value (the actual connection-string) but have non-equal names (keys).

PS:
More on my specific setup/why I do this:
https://stackoverflow.com/questions/28407510/neventstore-sagas-commands-and-not-losing-them

**Comment by andreabalducci** [comment]
posted on SO. Maybe it's better to have only one active thread.

**Comment by thomaseyde** [comment]
I think this issue is even worse. I stumbled on the same problem, see https://github.com/NEventStore/NEventStore.Persistence.SQL/issues/11.

It seems like this happens whenever we call `OpenStream()` on a background thread. In my case it looks like it's enough to delegate the work to a different thread, even the very first call gets the exception.

Is really the `DataReader` coupled to the `ConnectionScope`? It must somehow also be connected to the `streamId`, while the `bucketId` is ignored?

**Comment by alkampfergit** [comment]
Closed because inactive and because this problem can be solved with MARS = true.
